### PR TITLE
Add new job var CYLC_WORKFLOW_NAME_BASE.

### LINKED ISSUE #4977: CYLC_WORKFLOW_NAME: use the "flat" name
Many Cylc 7 workflows have been assuming a flat workflow ID (despite hierarchical registration going back to Cylc 6). With Cylc 8 hierarchical workflow IDs are now the default because run numbers are turned on by default.

To assist with the transition we added the `CYLC_WORKFLOW_NAME` environment variable (not to be mistaken with `CYLC_WORKFLOW_ID`) which is the ID with the run name/number stripped off.

Since https://github.com/cylc/cylc-flow/issues/4643 we have introduced another mechanism by which you can get hierarchical IDs "by default" in that we now transfer the directory structure from the src to the run dir.

Given its name you might expect the `CYLC_WORKFLOW_NAME` environment variable to reflect the "base name" of the workflow and not reflect its "path" either src or run. E.G for an ID of `a/b/run1` the "name" part might be considered to be `b` rather than `a/b`, especially since the src dir structure is likely to reflect project organisation.

**Question:**
* [ ] Should `CYLC_WORKFLOW_NAME` be "flat" or "hierarchical"?

**Pull requests welcome!**

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #4977 

As requested: did not add `CYLC_WORKFLOW_NAME_DIR`

Small compromise: I did not remove `CYLC_WORKFLOW_NAME` because  taken together it's immediately clear what the new variable is the basename of.

So now we have:
- `CYLC_WORKFLOW_ID` - e.g. `a/b/c/run1`
- `CYLC_WORKFLOW_NAME` - e.g. `a/b/c`
- `CYLC_WORKFLOW_NAME_BASE` - e.g. `c`

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [ ] No documentation update required.

### CODE REVIEW

**Comment by wxtim** [comment]
I've assigned myself to do a code review so that @dpmatthews doesn't have to, and so that @oliver-sanders  only has to do a second review.

File: tests/functional/restart/50-two-flows/flow.cylc
Comment (wxtim) [inline tests/functional/restart/50-two-flows/flow.cylc:25]:
```suggestion
               cylc trigger --flow=new --meta="cheese wizard" \
                   "$CYLC_WORKFLOW_ID//1/a"
```

File: tests/unit/test_job_file.py
Comment (wxtim) [inline tests/unit/test_job_file.py:283]: confusing as comments - recommend removal

```suggestion
```

**Review by wxtim — COMMENTED** [review]
- [x] Read
- [x] Checked out locally, and successfully run the changed tests.
- [ ] 1 broken test - suggested fix in comment.
- Thank you for getting rid of all those `${CYLC_WORKFLOW_NAMES}`.

**Review by wxtim — CHANGES_REQUESTED** [review]
See comments.

File: tests/functional/restart/50-two-flows/flow.cylc
Comment (hjoliver) [reply tests/functional/restart/50-two-flows/flow.cylc:25]: Ah, thanks @wxtim ; looks like I had an editor auto line break fail there after changing NAME to ID in the test.

File: tests/unit/test_job_file.py
Comment (hjoliver) [reply tests/unit/test_job_file.py:283]: Oops, I did not intend to leave those there - a left-over from my first attempt at adding the new variables to the test.

**Review by wxtim — APPROVED** [review]

**Comment by hjoliver** [comment]
Docs PR merged. One review OK for this.

**Pull request merged**
